src/sentry/similarity.py

```diff
@@ -0,0 +1,245 @@
+        where a similarity score of 1 is completely similar, and a similarity
+        score of 0 is completely dissimilar. The result sequence is ordered
+        from most similar to least similar. (For example, the search key itself
+        isn't filtered from the result and will always have a similarity of 1,
+        typically making it the first result.)
+        """
+        def fetch_bucket_frequencies(keys):
+            """Fetch the bucket frequencies for each band for each provided key."""
+            with self.cluster.map() as client:
+                responses = {
+                    key: map(
+                        lambda band: client.zrange(
+                            b'{}:{}:{}:{}:{}'.format(self.namespace, scope, self.BUCKET_FREQUENCY, band, key),
+                            0,
+                            -1,
+                            desc=True,
+                            withscores=True,
+                        ),
+                        range(len(self.bands)),
+                    ) for key in keys
+                }
+
+            result = {}
+            for key, promises in responses.items():
+                # Resolve each promise, and scale the number of observations
+                # for each bucket to [0,1] value (the proportion of items
+                # observed in that band that belong to the bucket for the key.)
+                result[key] = map(
+                    lambda promise: scale_to_total(
+                        dict(promise.value)
+                    ),
+                    promises,
+                )
+
+            return result
+
+        def fetch_candidates(signature):
+            """Fetch all the similar candidates for a given signature."""
+            with self.cluster.map() as client:
+                responses = map(
+                    lambda (band, buckets): map(
+                        lambda bucket: client.smembers(
+                            b'{}:{}:{}:{}:{}'.format(self.namespace, scope, self.BUCKET_MEMBERSHIP, band, bucket)
+                        ),
+                        buckets,
+                    ),
+                    enumerate(signature),
+                )
+
+            # Resolve all of the promises for each band and reduce them into a
+            # single set per band.
+            return map(
+                lambda band: reduce(
+                    lambda values, promise: values | promise.value,
+                    band,
+                    set(),
+                ),
+                responses,
+            )
+
+        target_frequencies = fetch_bucket_frequencies([key])[key]
+
+        # Flatten the results of each band into a single set. (In the future we
+        # might want to change this to only calculate the similarity for keys
+        # that show up in some threshold number of bands.)
+        candidates = reduce(
+            lambda total, band: total | band,
```

Comment:
Yeah, it's essentially the same thing although `operator.or_` is likely faster. (I didn't check, I might do this before merging.)